### PR TITLE
[bugfix] Fixes identify action on deactivated rules for QgsRuleBasedRenderer

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

This PR fixes an issue raised by @lbartoletti during its work on https://github.com/qgis/QGIS/pull/6657#issuecomment-375634020.

Actually, if a `ELSE` rule is defined and activated within the rule based renderer, then all others rules are valid even if they're deactivated. Then, an identify action can be done on invisible features.

A video is available, allowing to highlight the issue: https://github.com/lbartoletti/lbartoletti.github.io/tree/master/willRenderFeature.

A unit test has been added to test the underlying `willRenderFeature` method.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by pblottiere** [comment]
@nyalldawson Can you give me your opinion on this PR if you have some time? Thanks :)

**Comment by DelazJ** [comment]
@pblottiere I'm not sure but how close is this issue to https://issues.qgis.org/issues/18143?

**Comment by lbartoletti** [comment]
@pblottiere Tested and it's ok for me. Thanks!

This is not yours, but for other reviewers, a comment on [QgsNullSymbolRenderer](https://github.com/qgis/QGIS/pull/6679/files#diff-0f08397767c5914165b9a09fff136467R75) should be always false IMHO.

@DelazJ 👍

**Comment by pblottiere** [comment]
@Gustry 

Hi Etienne,

I think this PR fixes an issue you raised some time ago : https://issues.qgis.org/issues/13999

Before:

feature_count_bug

Now:

feature_count_fixed

**Comment by pblottiere** [comment]
@DelazJ 

> I'm not sure but how close is this issue to https://issues.qgis.org/issues/18143?

Mmm I'm not sure either. But this issue leads me to another one : https://issues.qgis.org/issues/13999

Thank you for that!

**Comment by DelazJ** [comment]
> But this issue leads me to another one : https://issues.qgis.org/issues/13999

Yes, also saw it when looking for the report. Wanted to point you to it but as it was mentioned in the first link, i just crossed my fingers and hoped you tackle both of them ;). Good it's gone; this was an annoying issue when willing to show feature count in legend. Thanks Paul.

File: src/core/symbology/qgsrulebasedrenderer.h [hunk 1]
```diff
@@ -343,7 +343,7 @@ class CORE_EXPORT QgsRuleBasedRenderer : public QgsFeatureRenderer
         QSet< QString > legendKeysForFeature( QgsFeature &feat, QgsRenderContext *context = nullptr );
 
         //! tell which rules will be used to render the feature
```
Comment (nyalldawson) [inline src/core/symbology/qgsrulebasedrenderer.h]: Maybe add an explanation of the `onlyActive` argument here too

**Review by nyalldawson — APPROVED** [review]
Looks great!

(You didn't happen to see any possible reason for the long-standing random crashes that the rule based renderer test experiences did you? That one has me totally stumped.)

**Comment by nyalldawson** [comment]
@lbartoletti 
>This is not yours, but for other reviewers, a comment on QgsNullSymbolRenderer should be always false IMHO.

Can we discuss that elsewhere? I personally think it's nice to be able to see selected features even with the null renderer.

File: src/core/symbology/qgsrulebasedrenderer.h (hunk 1 above)
Comment (pblottiere) [reply src/core/symbology/qgsrulebasedrenderer.h]: done

**Comment by pblottiere** [comment]
Travis is green, I'll merge this PR tomorrow if I have no more comments in the meantime.

**Comment by nyalldawson** [comment]
Great work @pblottiere!

**Comment by m-kuhn** [comment]
This causes a regression, I guess related to nested `ELSE` rules.

image

In such a scenario, the identify action will not discover visible features which are rendered by one of the nested rules within the `ELSE`.

**Pull request merged**